services/autoscaling/src/simcore_service_autoscaling/modules/instrumentation.py

```diff
@@ -0,0 +1,125 @@
+
+    def __post_init__(self) -> None:
+        self._active_nodes = Gauge(
+            "active_nodes",
+            "Number of EC2-backed docker nodes which are active and ready to run tasks",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._pending_nodes = Gauge(
+            "pending_nodes",
+            "Number of EC2-backed docker nodes which are active and NOT ready to run tasks",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._drained_nodes = Gauge(
+            "drained_nodes",
+            "Number of EC2-backed docker nodes which are drained",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._buffer_drained_nodes = Gauge(
+            "buffer_drained_nodes",
+            "Number of EC2-backed docker nodes which are drained and in buffer/reserve",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._pending_ec2s = Gauge(
+            "pending_ec2s",
+            "Number of EC2 instance not yet part of the cluster",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._disconnected_nodes = Gauge(
+            "disconnected_nodes",
+            "Number of docker node not backed by a running EC2 instance",
+        )
+        self._started_instances = Counter(
+            "started_instances_total",
+            "Number of EC2 instances that were started",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+        self._terminated_instances = Counter(
+            "terminated_ec2_instances_total",
+            "Number of EC2 instances that were terminated",
+            labelnames=EC2_INSTANCE_LABELS,
+        )
+
+    def update_from_cluster(self, cluster: Cluster) -> None:
+        _update_gauge(self._active_nodes, cluster.active_nodes)
+        _update_gauge(self._pending_nodes, cluster.pending_nodes)
+        _update_gauge(self._drained_nodes, cluster.drained_nodes)
+        _update_gauge(self._buffer_drained_nodes, cluster.reserve_drained_nodes)
+        _update_gauge(self._pending_ec2s, cluster.pending_ec2s)
+        self._disconnected_nodes.set(len(cluster.disconnected_nodes))
+
+    def instance_started(self, instance_type: str) -> None:
+        self._started_instances.labels(instance_type=instance_type).inc()
+
+    def instance_terminated(self, instance_type: str) -> None:
+        self._terminated_instances.labels(instance_type=instance_type).inc()
+
+
+def setup(app: FastAPI) -> None:
+    app_settings = get_application_settings(app)
+    if not app_settings.AUTOSCALING_PROMETHEUS_INSTRUMENTATION_ENABLED:
+        return
+
+    # NOTE: this must be setup before application startup
+    instrumentator = setup_rest_instrumentation(app)
+
+    async def on_startup() -> None:
+        app.state.instrumentation = AutoscalingInstrumentation(
+            registry=instrumentator.registry
+        )
+
+    async def on_shutdown() -> None:
+        if app.state.instrumentation:
+            ...
```

Comment:
what's the point of this if-statement?

Reply:
in case this would change. but it's not really necessary